src/Common/src/Interop/Unix/libcrypto/Interop.ERR.cs

```diff
@@ -17,13 +19,78 @@ internal static partial class libcrypto
         [DllImport(Libraries.LibCrypto, CharSet = CharSet.Ansi)]
         private static extern void ERR_error_string_n(uint e, StringBuilder buf, int len);
 
-        internal static string GetOpenSslErrorString()
+        private static string ERR_error_string_n(uint error)
         {
-            uint error = ERR_get_error();
             StringBuilder buf = new StringBuilder(1024);
-            
+
             ERR_error_string_n(error, buf, buf.Capacity);
             return buf.ToString();
         }
+
+        internal static Exception CreateOpenSslCryptographicException()
+        {
+            const uint ReasonMask = 0x0FFF;
+            const uint ERR_R_MALLOC_FAILURE = 64 | 1;
+
+            // The Windows cryptography library reports error codes through
+            // Marshal.GetLastWin32Error, which has a single value when the
+            // function exits, last writer wins.
+            //
+            // OpenSSL maintains an error queue. Calls to ERR_get_error read
+            // values out of the queue in the order that ERR_set_error wrote
+            // them. Nothing enforces that a single call into an OpenSSL
+            // function will guarantee at-most one error being set.
+            //
+            // In order to maintain parity in how error flows look between the
+            // Windows code and the OpenSSL-calling code, drain the queue
+            // whenever an Exception is desired, and report the exception
+            // related to the last value in the queue.
+            uint error = ERR_get_error();
+            uint lastRead = error;
+
+            // 0 (there's no named constant) is only returned when the calls
+            // to ERR_get_error exceed the calls to ERR_set_error.
+            while (lastRead != 0)
+            {
+                error = lastRead;
+                lastRead = ERR_get_error();
+            }
```

Comment:
If you wanted to have just one call to ERR_get_error, this could be rewritten like:
```C#
uint error = 0, lastRead;
while ((lastRead = ERR_get_error()) != 0)
{
    error = lastRead;
}
```